tests/conftest.py

```diff
@@ -57,10 +67,130 @@ def swarm_mode():
+        except (subprocess.CalledProcessError, FileNotFoundError):
+            logger.warning("Unable to parameterise the tests with podman")
+        else:
+            podman_client = DockerClient(client_call=["podman"])
+            podman_client.ctr_mgr = "podman"
+            ctr_clients.append(podman_client)
+    else:
+        unsupported_ctr_exes: List[str] = []
+        for exe in ctr_exes:
+            client = DockerClient(client_call=[exe])
+            # Check the client is available by running '<client> --version'.
+            try:
+                version_output = subprocess.check_output([exe, "--version"], text=True)
+            except (subprocess.CalledProcessError, FileNotFoundError):
+                unsupported_ctr_exes.append(exe)
+                continue
+            # Check the daemon is running (if required) with '<client> version'.
+            try:
+                subprocess.check_call([exe, "version"], stdout=subprocess.DEVNULL)
+            except subprocess.CalledProcessError:
+                unsupported_ctr_exes.append(exe)
+                continue
+
+            ctr_clients.append(client)
+
+            # Deduce whether docker or podman from version output.
+            if "docker" in version_output.lower():
+                client.ctr_mgr = "docker"
+            elif "podman" in version_output.lower():
+                client.ctr_mgr = "podman"
+            else:
+                logger.warning(
+                    f"Unrecognised container manager from '{exe} --version', "
+                    "assuming docker"
+                )
+                client.ctr_mgr = "docker"
+
+        if unsupported_ctr_exes:
+            raise Exception(
+                "'<ctr_exe> --version' failed for the following specified exes: "
+                + ",".join(unsupported_ctr_exes)
+            )
+
+    return ctr_clients
+
+
+def pytest_configure(config: pytest.Config) -> None:
+    # Filter warnings
+    if not pydantic.__version__.startswith("1"):
+        config.addinivalue_line(
+            "filterwarnings", "error::pydantic.PydanticDeprecatedSince20"
+        )
+    # Register markers
+    config.addinivalue_line(
+        "markers",
+        "ctr_mgr(MGR, MGR, ...): Container managers supported by the test",
+    )
+    # Determine supported container managers.
+    config.ctr_clients = _get_ctr_clients(config.getoption("--ctr-exe"))
+
+
+def pytest_generate_tests(metafunc: pytest.Metafunc) -> None:
+    if "ctr_client" in metafunc.fixturenames:
+        metafunc.parametrize(
+            "ctr_client",
+            metafunc.config.ctr_clients,
+            ids=[
+                client.client_config.client_call[0]
+                for client in metafunc.config.ctr_clients
+            ],
         )
+
+
+def pytest_collection_modifyitems(
+    config: pytest.Config, items: List[pytest.Item]
+) -> None:
+    # Find paramaterisations that don't make sense and remove them from
+    # the list of tests to be run.

```

Comment:
I'm not sold on this. Pytest already have a mechanism for this and it's the "SKIPPED" status. We should use it. I think it's important that the number of tests collected in the test suite never changes, but that tests can be skipped or deselected using pytest's buit-in features.

Reply:
Addressed in my main comment:
> There are actually two different cases where testcases might need to opt out from a runtime: the runtime does not support the functionality (e.g. docker does not support `docker pod` commands or `docker run --systemd false`); or POW does not yet support that functionality with the runtime. The reason I went with removing testcases was to cater for the former - in that situation the testcase simply doesn't make sense, so it seemed better for it not to show up as a SKIP which might imply missing functionality in POW. However, you're right that it would make more sense for the latter to be handled by skipping those, rather than removing them (which then serve as a kind of "TODO").

Reply:
I think we should carify what is expected from a pytest suite:
1) The number of collected items should always be the same, idependently of what is installed on the user's machine or the command line arguments. Here, it changes with the command line arguments, which does not satisfy this condition. This is mainly for visibility purposes and make it easy to reason about the test suite (the list of tests if fixed and the same for all contributors).
2) A test that should not be executed, based on command line arguments (`-k` or custom ones) should be `deselected`. It's what happens when we use `-k` or `-m`. Our custom command line arguments should have a similar effect since they have a similar purpose.
3) A test that can't be executed because a user doesn't have the necessary dependencies on his/her machine should be flagged as SKIPPED. For example, if someone doesn't have podman installed, all podman tests should be marked as SKIPPED, with the corresponding message explaining why.
4) A test that should pass but doesn't due to a bug known in the master branch should be marked as XFAIL, in pytest, it acts as a TODO and allows one person to know what is the expected behavior of the software, even if it doesn't work right now. It's a better documentation than a TODO comment. This can be useful is POW doesn't support something yet but should support it. It can also be used as a form of test-driven development.

If we respect those pytest rules, I don't really mind how it's implemented under the hood.

Reply:
This all seems reasonable, I'm happy to update to follow these rules.